crates/environ/src/func_environ.rs

```diff
@@ -494,6 +499,11 @@ impl lightbeam::ModuleContext for FuncEnvironment<'_> {
             .map(DefinedMemoryIndex::as_u32)
     }
 
+    fn vmctx_builtin_function(&self, func_index: u32) -> u32 {
+        self.offsets
+            .vmctx_builtin_function(BuiltinFunctionIndex::from_u32(func_index))
```

Comment:
Unfortunately the `BuiltinFunctionIndex` passed to this function and the `BuiltinFunctionIndex` used in the implementation of this function are not the same. I had to copy-paste the implementation to get the correct indices for the built-in functions and use `u32` for the types in the trait because otherwise I get cyclical dependencies - many of the types I would need to use are defined in `wasmtime-environ`, but Lightbeam is itself a dependency of `wasmtime-environ`.

Reply:
Ok. Could you add TODO comments mentioning that, as we discussed offline, the better approach here would be to start factoring out some of this into a separate crate that both Lightbeam and Cranelift could use?